File: 13.罗马数字转整数.py
#
# @lc app=leetcode.cn id=13 lang=python3
#
# [13] 罗马数字转整数
#

# @lc code=start
class Solution:
    def romanToInt(self, s: str) -> int:
        roma = {'I': 1,
                'V': 5,
                'X': 10,
                'L': 50,
                'C': 100,
                'D': 500,
                'M': 1000}
        res = 0
        if len(s) == 1:
            res = res = res + roma[s[0]]
        for i in range(1, len(s)):
            if roma[s[i-1]] < roma[s[i]]:
                res = res - roma[s[i-1]]
            else:
                res = res + roma[s[i-1]]
            if i == len(s)-1:
                res = res + roma[s[i]]
        return res
# @lc code=end

File: 13.罗马数字转整数.java

import java.util.*;

/*
 * @lc app=leetcode.cn id=13 lang=java
 *
 * [13] 罗马数字转整数
 */

// @lc code=start
class Solution {
    public int romanToInt(String s) {
        Map<Character, Integer> roma = new HashMap<>();
        roma.put('I', 1);
        roma.put('V', 5);
        roma.put('X', 10);
        roma.put('L', 50);
        roma.put('C', 100);
        roma.put('D', 500);
        roma.put('M', 1000);
        int res = 0;
        for(int i=0; i<s.length(); i++){
            if(i == s.length()-1){
                res = res + roma.get(s.charAt(i));
            }else{
                if(roma.get(s.charAt(i))<roma.get(s.charAt(i+1))){
                    res = res - roma.get(s.charAt(i));
                }
                else{
                    res = res + roma.get(s.charAt(i));
                }
            }
        }
        return res;
    }
}
// @lc code=end

